### PR TITLE
Parallelize checking for 10x performance win

### PR DESCRIPTION
**Pull request by aatifsyed**
## What does this PR accomplish?
- Improves checking times by ~10x

comparison with `cargo-spellcheck 0.12.4` by running `cargo spellcheck` on https://github.com/rust-lang/rust/tree/22e9fe644ea710eec50cb0aabcae7fa8dd9fd675

(see below note for why I've done `-j $(nproc)`
```console
$ hyperfine "cargo spellcheck -j $(nproc)" "~/code/cargo-spellcheck/target/release/cargo-spellcheck -j $(nproc)" --export-markdown hyperfine.md 
Benchmark 1: cargo spellcheck -j 16
  Time (mean ± σ):     356.515 s ± 11.933 s    [User: 409.724 s, System: 155.652 s]
  Range (min … max):   340.755 s … 374.962 s    10 runs

Benchmark 2: ~/code/cargo-spellcheck/target/release/cargo-spellcheck -j 16
  Time (mean ± σ):     32.953 s ±  2.922 s    [User: 353.613 s, System: 13.498 s]
  Range (min … max):   26.296 s … 36.698 s    10 runs

Summary
  '~/code/cargo-spellcheck/target/release/cargo-spellcheck -j 16' ran
   10.82 ± 1.03 times faster than 'cargo spellcheck -j 16'
```

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `cargo spellcheck -j 16` | 356.515 ± 11.933 | 340.755 | 374.962 | 10.82 ± 1.03 |
| `~/code/cargo-spellcheck/target/release/cargo-spellcheck -j 16` | 32.953 ± 2.922 | 26.296 | 36.698 | 1.00 |

 * 🩹 Bug Fix
 * 🦚 Feature
 

Closes # .

## Changes proposed by this PR:
- Parallelise running the checkers

## Notes to reviewer:
Not sure why the previous checking code was single threaded - it was definitely _attempting_ to be parallel, but misunderstood the stream APIs.

- There are probably a bunch more perf improvements to make - suspect the next bottleneck is document collection.
- Would be nice to use `tracing`, and have better (parallel) profiling tools.
- Not sure why we use the number of physical cpus - hunspell is using only half the threads it can. We should use `num_cpus::get` instead of `num_cpus::get_physical` (or `std::thread::available_parallelism`).
- clippy is _very unhappy_ with `doc-chunks`
- lots of methods on `Document` should probably be removed

## 📜 Checklist

 * [x] Works on the `./demo` sub directory
 * [x] Test coverage is excellent and passes
   No change
 * [x] Documentation is thorough
   No change

### CODE REVIEW

File: src/action/mod.rs [hunk 1]
```diff
@@ -407,33 +407,28 @@ impl Action {
...
+                checkers.check(&origin, &chunks).map(|suggestions| {
+                    let path = origin.as_path();
+                    let n = suggestions.len();
+                    match suggestions.is_empty() {
+                        true => log::info!("✅ {}", path.display()),
+                        false => log::info!("❌ {} : {}", path.display(), n),
+                    };
+                    for suggestion in suggestions {
+                        println!("{}", suggestion);
+                    }
```
Comment (drahnr) [inline src/action/mod.rs:423]: Wouldn't the log output be scrambled/interleaved with one another?

**Comment by drahnr** [comment]
My results appear to be very different on 40k loc project using the following command:

```
cargo b --release; hyperfine --show-output -r 2 "./target/release/cargo-spellcheck check -j16 --code 0 ../secret/project"
```

Current main branch
```
# git 8e4cfcce7f8cf3984664feba44e90dabe9d9fcf1
Benchmark 1: ./target/release/cargo-spellcheck check -j16 --code 0 ../gensyn/node
  Time (mean ± σ):     191.1 ms ±   5.7 ms    [User: 246.0 ms, System: 55.1 ms]
  Range (min … max):   187.1 ms … 195.2 ms    2 runs
```

```
# git 6c50e07efa303aa37fff73be1d7b187b36dc3e22
Benchmark 1: ./target/release/cargo-spellcheck check -j16 --code 0 ../gensyn/node
  Time (mean ± σ):     184.6 ms ±   3.2 ms    [User: 236.8 ms, System: 58.1 ms]
  Range (min … max):   182.4 ms … 186.9 ms    2 runs
```

which aligns with the CPU utilization I see - it's only one core regardless.

**Comment by drahnr** [comment]
```
hyperfine --show-output "./cargo-spellcheck-8e4cfcce7f8cf3984664feba44e90dabe9d9fcf1 -j 16 --code 0 ../gensyn/node" "./cargo-spellcheck-6c50e07efa303aa37fff73be1d7b187b36dc3e22 -j 16 --code 0 ../gensyn/node" --export-markdown hyperfine.md 
Benchmark 1: ./cargo-spellcheck-8e4cfcce7f8cf3984664feba44e90dabe9d9fcf1 -j 16 --code 0 ../gensyn/node
  Time (mean ± σ):     180.1 ms ±   6.3 ms    [User: 238.3 ms, System: 59.6 ms]
  Range (min … max):   170.1 ms … 191.2 ms    15 runs
 
Benchmark 2: ./cargo-spellcheck-6c50e07efa303aa37fff73be1d7b187b36dc3e22 -j 16 --code 0 ../x
  Time (mean ± σ):     178.5 ms ±   6.7 ms    [User: 235.5 ms, System: 57.5 ms]
  Range (min … max):   171.0 ms … 191.0 ms    15 runs
 
Summary
  ./cargo-spellcheck-6c50e07efa303aa37fff73be1d7b187b36dc3e22 -j 16 --code 0 ../gensyn/node ran
    1.01 ± 0.05 times faster than ./cargo-spellcheck-8e4cfcce7f8cf3984664feba44e90dabe9d9fcf1 -j 16 --code 0 ../x
    ```

File: src/action/mod.rs (hunk 1 above)
Comment (drahnr) [reply src/action/mod.rs:423]: It doesn't practically.. :monocle_face: that is a bit fishy

**Comment by drahnr** [comment]
@aatifsyed 👋

**Comment by aatifsyed** [comment]
Interesting - just to clarify my understanding, on _this_ branch you're seeing only a single core used?

**Comment by drahnr** [comment]
> Interesting - just to clarify my understanding, on _this_ branch you're seeing only a single core used?

Yes, exactly. The pattern differs, but there is only ever one CPU used at 100% with this branch. I suspected it being IO bound or a lock in `stdout` (hypothesis).

**Comment by aatifsyed** [comment]
That's very surprising.

I've prepared tracing on two branches for you:
https://github.com/aatifsyed/cargo-spellcheck/tree/parallel-traced (my code)
https://github.com/aatifsyed/cargo-spellcheck/tree/single-thread-traced (pre-existing code)

Running each will give you a `trace-<unix_seconds>.json` that you can view at https://ui.perfetto.dev, or chome://tracing in a chromium browser.

Here are the traces on my machine (for a bare `cargo spellcheck` invocation), on https://github.com/ChainSafe/forest/:
```
$ loc
--------------------------------------------------------------------------------
 Language             Files        Lines        Blank      Comment         Code
--------------------------------------------------------------------------------
 Rust                   313        51683         5975         5028        40680
 Markdown                27         4669          823            0         3846
...
```
parallel-traced (my code)
image
you can see the multiple threads of execution
[trace-parallel.json.txt](https://github.com/drahnr/cargo-spellcheck/files/12064683/trace-parallel.json.txt)


single-thread-traced (pre-existing code)
image
you can see that there only a single thread
[trace-single-threaded.json.txt](https://github.com/drahnr/cargo-spellcheck/files/12064684/trace-single-threaded.json.txt)

**Comment by aatifsyed** [comment]
>  I suspected it being IO bound or a lock in stdout (hypothesis).

I'd be surpised if this was the case!

File: src/action/mod.rs (hunk 1 above)
Comment (aatifsyed) [reply src/action/mod.rs:423]: This is a fair concern - we should probably fan the work out to threads, and only present on the main thread

**Comment by drahnr** [comment]
[parallel.json.txt](https://github.com/drahnr/cargo-spellcheck/files/12067145/parallel.json.txt)
[vanilla.json.txt](https://github.com/drahnr/cargo-spellcheck/files/12067146/vanilla.json.txt)

I am on fedora 38 AMD 64 with a 16 core processor.

**Comment by drahnr** [comment]
[forest-parallel.json.txt](https://github.com/drahnr/cargo-spellcheck/files/12067310/forest-parallel.json.txt)
[forest-vanilla.json.txt](https://github.com/drahnr/cargo-spellcheck/files/12067311/forest-vanilla.json.txt)

When running on `forest`, I indeed see the ~10x speedup.

---

I'll have to debug why this is not the case for the private repo, but I don't see any obstacles to merging this as is.

**Review by drahnr — APPROVED** [review]

**Comment by drahnr** [comment]
Alright, I think I figured the root cause, it's rather unintuitive cli flag interpretation.

**Pull request merged**